contracts/EpochTokenLocker.sol

```diff
@@ -90,57 +104,61 @@ contract EpochTokenLocker {
      * view functions
      */
     function getPendingDepositAmount(address user, address token) public view returns(uint) {
-        return balanceStates[user][token].pendingDeposits.amount;
+        return balanceStates[user].balances[token].pendingDeposits.amount;
     }
 
     function getPendingDepositBatchNumber(address user, address token) public view returns(uint) {
-        return balanceStates[user][token].pendingDeposits.stateIndex;
+        return balanceStates[user].balances[token].pendingDeposits.stateIndex;
     }
 
     function getPendingWithdrawAmount(address user, address token) public view returns(uint) {
-        return balanceStates[user][token].pendingWithdraws.amount;
+        return balanceStates[user].balances[token].pendingWithdraws.amount;
     }
 
     function getPendingWithdrawBatchNumber(address user, address token) public view returns(uint) {
-        return balanceStates[user][token].pendingWithdraws.stateIndex;
+        return balanceStates[user].balances[token].pendingWithdraws.stateIndex;
     }
 
     function getCurrentStateIndex() public view returns(uint32) {
         return uint32(now / BATCH_TIME);
     }
 
     function getBalance(address user, address token) public view returns(uint256) {
-        uint balance = balanceStates[user][token].balance;
-        if (balanceStates[user][token].pendingDeposits.stateIndex < getCurrentStateIndex()) {
-            balance = balance.add(balanceStates[user][token].pendingDeposits.amount);
+        uint balance = balanceStates[user].balances[token].balance;
+        if (balanceStates[user].balances[token].pendingDeposits.stateIndex < getCurrentStateIndex()) {
+            balance = balance.add(balanceStates[user].balances[token].pendingDeposits.amount);
         }
-        if (balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentStateIndex()) {
-            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        if (balanceStates[user].balances[token].pendingWithdraws.stateIndex < getCurrentStateIndex()) {
+            balance -= Math.min(balanceStates[user].balances[token].pendingWithdraws.amount, balance);
         }
         return balance;
     }
 
+    function getAllAccounts() public view returns (address[] memory) {
```

Comment:
I think this has an scalability issue, right? 

Would it make sense to implementing like:
```solidity
function getAllAccounts(uint startIndex, uint count) public view returns (address[] memory) {
     // if count = 0 or greater than size, return the array
     // else return a slice of it
}
```

Reply:
Is this is an issue for view functions?

Reply:
I think there's a limit. Not sure what is it, any ways, if the list can be big, you might be inefficient transfering all the accounts.

Imagine an interface where you display the 50 first users and then you allow to iterate into next page. You'll transfer all the users just to display the first page.

Reply:
This is beginning to sound like a scalability issue for an external DB? Although I though the PoC was not concerned with scalable aspects of the protocol.

Reply:
> Imagine an interface where you display the 50 first users and then you allow to iterate into next page. 

I'd assume an interface to do some kind of sorting (e.g. by price, token whatsoever). None of this is possible with a query like this. This query is at this point primarily designed for the driver to get all the information it needs to pass to the solver. I don't think pagination makes much sense there.

Reply:
Okok